lib/joystick.js

```diff
@@ -11,149 +125,124 @@ var Board = require("../lib/board.js"),
-    y: 0
-  };
-
-  this.normalized = {
-    x: 0,
-    y: 0
-  };
-
-  this.raw = {
-    x: 0,
-    y: 0
-  };
-
-
-  // TODO: calculate joystick direction based on X/Y values
-  this.direction = "";
+  if (opts.controller && typeof opts.controller === "string") {
+    controller = Controllers[opts.controller.toUpperCase()];
+  } else {
+    controller = opts.controller;
+  }
 
-  this.pins.forEach(function(pin) {
-    // Set the pin to input mode
-    this.io.pinMode(pin, this.mode);
+  if (controller == null) {
+    controller = Controllers["ANALOG"];
+  }
 
-    // Arduino 0-1023  512
-    this.io.analogRead(pin, function(data) {
+  Object.defineProperties(this, controller);
 
-      var val = data / 1023;
+  if (!this.toAxis) {
+    this.toAxis = opts.toAxis || function(raw) { return raw; };
+  }
 
-      // register X axis, L/R (A0)
-      if (pin === 1) {
-        this.axis.x = val;
-        this.raw.x = data;
-        // register Y axis, U/D (A1)
-      } else {
-        this.axis.y = val;
-        this.raw.y = data;
+  this.threshold = opts.threshold === undefined ? 0.01 : opts.threshold;
+
+  state.x.zeroV = opts.zeroV === undefined ? 0 : (opts.zeroV.x || 0);
+  state.y.zeroV = opts.zeroV === undefined ? 0 : (opts.zeroV.y || 0);
+
+  state.x.invert = opts.invertX || opts.invert || false;
+  state.y.invert = opts.invertY || opts.invert || false;
+
+  priv.set(this, state);
+
+  if (typeof this.initialize === "function") {
+    this.initialize(opts, function(data) {
+      var isChange = false;
+      var computed = {
+        x: null,
+        y: null
+      };
+
+      Object.keys(data).forEach(function(axis) {
+        var value = data[axis];
+        var sensor = state[axis];
+
+        // Set the internal ADC reading value...
+        sensor.value = value - sensor.zeroV;
+        // ... Get the computed axis value.
+        computed[axis] = this[axis];
+
+        var absAxis = Math.abs(computed[axis]);
+        var absPAxis = Math.abs(sensor.previous);
+
+        if ((absAxis < absPAxis - this.threshold) ||
+            (absAxis > absPAxis + this.threshold)) {
+          sensor.previous = computed[axis];
+          isChange = true;
+        }
+      }, this);
+
+      this.emit("data", {
+        x: computed.x,
+        y: computed.y
+      });
+
+      if (isChange) {
+        this.emit("change", {
+          x: computed.x,
+          y: computed.y
+        });
       }
     }.bind(this));
-  }.bind(this));
-
-
-  // Throttle event emitter
-  setInterval(function() {
-
-    this.emit("axismove", err, new Date());
-
-  }.bind(this), this.freq);
+  }
 
   Object.defineProperties(this, {
-    /**
-     * [read-only] Calculated magnitude value
-     * @name magnitude
-     * @property
-     * @type Number
-     */
-    magnitude: {
-      get: function() {
-        return Math.hypot(this.axis.x, this.axis.y);
+    hasAxis: {
+      value: function(axis) {
+        return state[axis] ? typeof state[axis].value !== "undefined" : false;
```

Comment:
Won't this always return `true` for `"x"` and `"y"` and `false` for any other value?

It looks like this was designed to handle a three axis joystick as well, but there's no way to actually read from the third axis. Similarly, I can power my Parallax joystick such that only one axis will work, but I wouldn't be able to use this component without providing two pins to read from since the data events are blocked on reading values from both axes.